### DISCUSSION TITLE
[Feature request] DataSourceManager.initDataSource 时提升效率

### DISCUSSION THREAD

**Discussion by geekberryonekey** [midwayjs/midway#4003] [Ideas]
### The feature, motivation（功能、动机）

目前在使用 dataSource 链接多个数据源时, 每个数据源的链接是串行执行的. 在数据源数量较多或者部分链接较慢时, 会导致启动时间非常长

### Alternatives（替代方案）

https://github.com/midwayjs/midway/blob/0edf6c16f2e451473dc2502966bda96cf10e589f/packages/core/src/common/dataSourceManager.ts#L61

将
```
for () {
  await ...
}
```

改为

```
await Promise.all([])
```

结构

### Additional context（其他上下文）

@midwayjs/core 3.15.0

**Comment by geekberryonekey** [comment]
请问这个问题有计划解决吗?

**Reply by czy88840616** [reply]
v3 不动，v4 已加。

**Comment by geekberryonekey** [comment]
请问是否可以考虑给 V3 升级一下这个功能?
现在我们有越来越多的的数据源要连接, 新建容器的启动时间越来越长, 对扩容和可用性有造成了影响.
但是由于依赖 V3 的项目数量众多, 且项目功能越发复杂, 暂时不能及时升级到 V4.

以下是我们, 本地已长期验证过的改写方式:

/node_modules/@midwayjs/core/dist/common/dataSourceManager.js
```js
// L47: for (const dataSourceName in dataSourceConfig.dataSource) {
await Promise.all(Object.keys(dataSourceConfig.dataSource).map(async dataSourceName => {

// L78: }
}));
```

**Comment by czy88840616** [comment]
这个能力在 v4 里已经做了，当前 `DataSourceManager.initDataSource` 支持 `concurrent`，`@midwayjs/typeorm` 初始化时也已经传了 `concurrent: true`，所以 v4 下多个 typeorm dataSource 会并发初始化。

V3 是否 backport 需要再评估一下兼容性。主要风险不是代码改动大小，而是串行初始化变并发后，用户项目里如果依赖 dataSource 的初始化顺序，或者某些组件在初始化期间有共享副作用，行为可能会变化。

如果只在 V3 的 typeorm 场景下增加一个可选配置，并保持默认串行，会更稳一些。你们本地已经验证过的话，可以贴一下完整改法，或者直接提一个 PR：

- 默认仍保持 V3 现有串行行为；
- 新增可选开关启用并发；
- 补一个多 dataSource 并发初始化的单测。

这样比较容易合到 V3 维护分支里。

**Comment by geekberryonekey** [comment]
v3 按照 [3.20.24 的修改](https://github.com/midwayjs/midway/commit/fd8efe6f3fd5fa257979a97c420602fc5bde2e2b#diff-81b8343103c1a63b96fa7a483b7fd23e5ae43167991382e5d7ec846b62ce600aR459) 在 config.core.concurrentDataSourceInitialization 设置为 true 可以并行了.

**Comment by czy88840616** [comment]
是的，v3 这块已经按这个方向 backport 了。

`3.20.24` 之后可以通过：

```ts
// config.default.ts
export default {
  core: {
    concurrentDataSourceInitialization: true,
  },
};
```

来打开 dataSource 并行初始化。这里默认仍然保持串行，主要是为了兼容 v3 里可能依赖 dataSource 初始化顺序的老项目；明确确认多个数据源之间没有顺序依赖后，再开启这个配置会更稳。

另外这个开关会传到 core 的 `DataSourceManager.initDataSource(..., { concurrent })`，并行处理 entity glob 和 dataSource 创建，所以你们这种多数据源启动慢的场景可以优先升级到 `3.20.24+` 后验证启动耗时。
